### PR TITLE
Refactor InputTransform and DataModule

### PR DESCRIPTION
**Pull request by karthikrangasai**
## What does this PR do?

Resolves discussion from #1166 

At present, the `InputTransform` for every `DataModule` is being passed for every stage even though `InputTransform` has specific methods to differentiate the stage at which the certain transform runs. This also creates 4 different instances of the class in each input (train, val, test, predict), with different methods to be run.

This PR aims to change the aforementioned by making the `DataModule` as the owner of the `InputTransform` because all the class does it to generate `collate_fn` for the dataloaders and the implementation for the `on_after_batch_transfer` callback.

Thus a single instance of the `InputTransform` class, present in the DataModule, can resolve the required `Callable`s for every stage and the appropriate dataloader `collate_fn` and `on_after_batch_transfer` functions are created in the `DataLoader`'s `__init__` method.

This also relieves the `Input` class from having to take care of the `InputTransform`.

TL;DR
### Previous API
```python
dm = XYZTask_DataModule.from_xyz(
    train_file=train_file,
    val_file=val_file,
    test_file=test_file,
    predict_file=predict_file,
    train_transform=InputTransform,
    val_transform=InputTransform,
    test_transform=InputTransform,
    predict_transform=InputTransform,
    transform_kwargs=transform_kwargs,
)

# Implementation
class XYZTask_DataModule(DataModule):
    
    @classmethod
    def from_xyz(
        cls,
        train_file=train_file,
        val_file=val_file,
        test_file=test_file,
        predict_file=predict_file,
        train_transform=InputTransform,
        val_transform=InputTransform,
        test_transform=InputTransform,
        predict_transform=InputTransform,
        transform_kwargs=transform_kwargs,
        input_cls=Input,
    ):
        return cls(
            input_cls(RunningStage.TRAINING, train_file, transform=train_transform, **transform_kwargs),
            input_cls(RunningStage.VALIDATING, val_file, transform=val_transform, **transform_kwargs),
            input_cls(RunningStage.TESTING, test_file, transform=test_transform, **transform_kwargs),
            input_cls(RunningStage.PREDICTING, predict_file, transform=predict_transform, **transform_kwargs),
        )
```

### New API
```python
dm = XYZTask_DataModule.from_xyz(
    train_file=train_file,
    val_file=val_file,
    test_file=test_file,
    predict_file=predict_file,
    transform=InputTransform,
    transform_kwargs=transform_kwargs,
)

# Implementation
class XYZTask_DataModule(DataModule):
    
    @classmethod
    def from_xyz(
        cls,
        train_file=train_file,
        val_file=val_file,
        test_file=test_file,
        predict_file=predict_file,
        transform=InputTransform,
        transform_kwargs=transform_kwargs,
        input_cls=Input,
    ):
        return cls(
            input_cls(RunningStage.TRAINING, train_file),
            input_cls(RunningStage.TRAINING, val_file),
            input_cls(RunningStage.TRAINING, test_file),
            input_cls(RunningStage.TRAINING, predict_file),
            transform=transform,
            **transform_kwargs
        )

```

## Before submitting
- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the **[contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md)**, Pull Request section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes?
- [ ] Did you write any **new necessary tests**? [not needed for typos/docs]
- [ ] Did you verify **new and existing tests pass** locally with your changes?
- [ ] If you made a notable change (that affects users), did you **update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)**?

## PR review
 - [x] Is this pull request **ready for review**? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by karthikrangasai** [comment]
Hello all,

I have updated the PR for almost all tasks.

I need some help for:

1. `ServeInput` and serve related parts.
2. Pointcloud tasks retrieve their `collate_fn` when model is instantiated. The updated `_collate_fn` method in `InputTransform` class accepts an extra argument (stage: RunningStage) and I wasn't sure of how to set that up in a cleaner manner.
3. Same issue as above with the `image.ObjectDetectionTask` which fails for the `test_init_train` testing function. The extra argument is causing an error. Need some help in updating/changing this test.

Thanks

File: flash/core/data/data_module.py [hunk 1]
```diff
@@ -467,6 +489,55 @@ def data_fetcher(self) -> BaseDataFetcher:
...
+        Override with your custom one.
+        """
+        return create_or_configure_input_transform(
+            transform=transform,
+            transform_kwargs=transform_kwargs,
+            input_transforms_registry=DataModule.input_transforms_registry,
+        )
+
+    @classmethod
+    def register_input_transform(
```
Comment (ethanwharris) [inline flash/core/data/data_module.py:520]: Do we need these?

File: flash/audio/speech_recognition/input_transform.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+from flash.core.utilities.imports import _AUDIO_AVAILABLE
+
+if _AUDIO_AVAILABLE:
+    from transformers import AutoProcessor
+else:
+    AutoProcessor = object
+
+
+@dataclass
+class SpeechRecognitionInputCollateTransform(InputTransform):
```
[the remark this answers is not in the document]
Comment (karthikrangasai) [reply flash/audio/speech_recognition/input_transform.py:28]: I was initially planning to make it a separate PR, but the tests for these tasks were failing. So I ended up implementing them all here.

File: flash/tabular/forecasting/input_transform.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+from flash.core.utilities.imports import _FORECASTING_AVAILABLE
+
+if _FORECASTING_AVAILABLE:
+    from pytorch_forecasting import TimeSeriesDataSet
+else:
+    TimeSeriesDataSet = object
+
+
+@dataclass
+class TabularForecastingInputCollateTransform(InputTransform):
```
[the remark this answers is not in the document]
Comment (karthikrangasai) [reply flash/tabular/forecasting/input_transform.py:35]: I have taken the implementation from their library only.

The issue is that the collate function needs another parameter as I had mentioned above.
This was causing an issue with the tests. Need some workaround here.

File: tests/text/question_answering/test_model.py [hunk 4]
```diff
@@ -46,13 +54,13 @@ def __len__(self) -> int:
...
 
 
-@pytest.mark.skipif(os.name == "nt", reason="Huggingface timing out on Windows")
-@pytest.mark.skipif(not _TEXT_TESTING, reason="text libraries aren't installed.")
-def test_init_train(tmpdir):
-    model = QuestionAnsweringTask(TEST_BACKBONE)
-    train_dl = torch.utils.data.DataLoader(DummyDataset())
-    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True)
-    trainer.fit(model, train_dl)
+# @pytest.mark.skipif(os.name == "nt", reason="Huggingface timing out on Windows")
```
[the remark this answers is not in the document]
Comment (karthikrangasai) [reply tests/text/question_answering/test_model.py:57]: Forgot to mention in the discussions point.

The new DummyDataset was not compatible with the training and raised some tokenization errors.
Wanted to get the team's thought on how to improve that test.

File: flash/text/question_answering/input_transform.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+if _TRANSFORMERS_AVAILABLE:
+    from transformers import DataCollatorWithPadding, default_data_collator
+else:
+    DataCollatorWithPadding = object
+    default_data_collator = object
+
+
+@dataclass(unsafe_hash=True)
+class QuestionAnsweringInputTransform(TransformersTextInputTransform):
```
[the remark this answers is not in the document]
Comment (karthikrangasai) [reply flash/text/question_answering/input_transform.py:36]: Every `from_*` method of the DataModule runs a full training sequence which loads all the data and transforms it. So I guess it is being tested properly. I can add a specific test to test it further.

File: flash/text/question_answering/model.py [hunk 6]
```diff
@@ -278,25 +262,22 @@ def forward(self, batch: Any) -> Any:
...
-        predicted_answers = [generated_tokens[example["example_id"]] for example in batch]
+    def compute_metrics(self, generated_answers, batch):
+        predicted_answers = [generated_answers[example["example_id"]] for example in batch]
         target_answers = [
             example["answer"]["text"][0] if len(example["answer"]["text"]) > 0 else "" for example in batch
         ]
         return self.rouge(predicted_answers, target_answers)
 
+    def training_step(self, batch: Any, batch_idx: int) -> Tensor:
+        self.common_step("train", batch)
```
[the remark this answers is not in the document]
Comment (karthikrangasai) [reply flash/text/question_answering/model.py:279]: My bad. Will fix it.

File: flash/audio/speech_recognition/input_transform.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/audio/speech_recognition/input_transform.py:28]: I guess we just need to resolve the collate function from the models correctly (or there are issues with it). The main reason to leave this for the future is that I'm not sure this is where this functionality should end up. To me it's weird that the transform can include collation since that's not really a transform. I also think we should avoid a situation where users have to provide the backbone in two places (since only the model should know the backbone really).

File: flash_examples/flash_components/custom_data_loading.py [hunk 7]
```diff
@@ -38,23 +38,25 @@
...
 # The folder paths are independent and when loading the order of folder.                    #
 # would determine the classification label.                                                 #
-# Note: This is simple enough to show you the flexibility of the Flash API.                 #
+# NOTE: This is simple enough to show you the flexibility of the Flash API.                 #
 #############################################################################################
 
 
 #############################################################################################
-#                         Step 1 / 2: Implement a Input                                     #
+#                         Step 1 / 3: Implement a Input                                     #
```
Comment (krshrimali) [inline flash_examples/flash_components/custom_data_loading.py:46]:
```suggestion
#                         Step 1 / 3: Implement an Input                                    #
```

File: flash_examples/flash_components/custom_data_loading.py [hunk 8]
```diff
@@ -92,11 +94,22 @@ def predict_load_data(self, predict_folder: str) -> List[Dict[DataKeys, Any]]:
 
 
 #############################################################################################
-#                   Step 2 / 2: [optional] Implement a InputTransform                       #
+#                   Step 2 / 3: [optional] Implement a InputTransform                       #
```
Comment (krshrimali) [inline flash_examples/flash_components/custom_data_loading.py:97]:
```suggestion
#                   Step 2 / 3: [optional] Implement an InputTransform                      #
```

File: flash_examples/flash_components/custom_data_loading.py [hunk 9]
```diff
@@ -92,11 +94,22 @@ def predict_load_data(self, predict_folder: str) -> List[Dict[DataKeys, Any]]:
...
 #                                                                                           #
 # A `InputTransform` is a state-aware (c.f training, validating, testing and predicting)    #
```
Comment (krshrimali) [inline flash_examples/flash_components/custom_data_loading.py:99]:
```suggestion
# An `InputTransform` is a state-aware (c.f training, validating, testing and predicting)   #
```

File: flash_examples/flash_components/custom_data_loading.py [hunk 10]
```diff
@@ -105,7 +118,7 @@ class BaseImageInputTransform(InputTransform):
 
     image_size: Tuple[int, int] = (224, 224)
 
-    def input_per_sample_transform(self) -> Any:
+    def input_per_sample_transform(self) -> Callable:
```
Comment (krshrimali) [inline flash_examples/flash_components/custom_data_loading.py:121]: Nice attention to detail :)

**Review by krshrimali — APPROVED** [review]
Awesome work, @karthikrangasai - lots of important refactoring done in this PR. Thank you!

We are getting there, just left a few comments (minor nits, and a few questions for my knowledge). Just taking a note for the future, we would probably need to update the examples as well since the API has changed. (cc: @ethanwharris @Borda)

File: flash/core/data/io/input_transform.py [hunk 11]
```diff
@@ -82,33 +83,30 @@ def __repr__(self):
 
 
 @dataclass
-class InputTransform(Properties):
+class _InputTransformPerStage:
+    collate_in_worker_from_transform: Optional[bool] = None
+    transforms: Optional[Dict[str, Callable]] = None
```
[the remark this answers is not in the document]
Comment (karthikrangasai) [reply flash/core/data/io/input_transform.py:88]: I called it `transforms` because it hold one for every hook. Also, the typing makes it clear.
Should I be changing it ?

File: flash/core/data/io/input_transform.py [hunk 12]
```diff
@@ -1170,15 +1183,20 @@ def __str__(self) -> str:
...
-    """This utility is used to create the 2 `_InputTransformProcessorV2` objects which contain the transforms used
-    as the DataLoader `collate_fn` and the DataModule `on_after_batch_transfer` hook."""
+def __make_collates(input_transform: "InputTransform", on_device: bool, collate: Callable) -> Tuple[Callable, Callable]:
+    if on_device:
+        return input_transform._identity, collate
+    return collate, input_transform._identity
+
+
+def __configure_worker_and_device_collate_fn(
+    running_stage: RunningStage, input_transform: "InputTransform"
```
[the remark this answers is not in the document]
Comment (karthikrangasai) [reply flash/core/data/io/input_transform.py:1193]: Since it was in the same file, type reference might not load (happens in the case when we have circular imports, not this maybe). So I put them just to be safe.

[the remark this answers is not in the document]
Comment (krshrimali) [reply flash/core/data/io/input_transform.py]: Oh, makes sense. I missed that. Thanks!

**Review by ethanwharris — APPROVED** [review]
Awesome, LGTM 😃

**Comment by tchaton** [comment]
Awesome work @karthikrangasai !
